File: mcp/tools/mcp-use/libraries/python/examples/code_mode_example.py
"""
Code Mode Example - Using MCP Tools via Code Execution

This example demonstrates how AI agents can use MCP tools through code execution mode,
which enables more efficient context usage and data processing compared to
direct tool calls.

Based on Anthropic's research: https://www.anthropic.com/engineering/code-execution-with-mcp
"""

import asyncio

from langchain_anthropic import ChatAnthropic

from mcp_use import MCPAgent, MCPClient
from mcp_use.client.prompts import CODE_MODE_AGENT_PROMPT

# Example configuration with a simple MCP server
# You can replace this with your own server configuration
config = {
    "mcpServers": {
        "filesystem": {
            "command": "npx",
            "args": ["-y", "@modelcontextprotocol/server-filesystem", "."],
        }
    }
}


async def main():
    """Example 5: AI Agent using code mode (requires OpenAI API key)."""

    client = MCPClient(config=config, code_mode=True)

    # Create LLM
    llm = ChatAnthropic(model="claude-haiku-4-5-20251001")

    # Create agent with code mode instructions
    agent = MCPAgent(
        llm=llm,
        client=client,
        system_prompt=CODE_MODE_AGENT_PROMPT,
        max_steps=50,
        pretty_print=True,
    )

    # Example query
    query = """ Please list all the files in the current folder."""

    async for _ in agent.stream_events(query):
        pass


if __name__ == "__main__":
    asyncio.run(main())

File: mcp/tools/mcp-use/libraries/typescript/packages/agent/examples/code-mode/code_mode_example.ts
/**
 * Code mode — MCP tools via in-process code execution.
 *
 * Run: pnpm exec tsx examples/code-mode/code_mode_example.ts
 * Requires: ANTHROPIC_API_KEY
 */

import { ChatAnthropic } from "@langchain/anthropic";
import fs from "node:fs";
import os from "node:os";
import path from "node:path";
import { MCPAgent, PROMPTS } from "@mcp-use/agent/langchain";
import { MCPClient } from "@mcp-use/client";
import { ANTHROPIC_MODEL } from "../_shared.js";

const tempDir = fs.mkdtempSync(path.join(os.tmpdir(), "mcp-use-code-mode-"));
const filePath = path.join(tempDir, "test.txt");
fs.writeFileSync(filePath, "Hello, world!");
console.log("Temp dir:", tempDir);

async function main() {
  if (!process.env.ANTHROPIC_API_KEY) {
    console.error("Missing ANTHROPIC_API_KEY");
    process.exit(1);
  }

  const client = new MCPClient(
    {
      mcpServers: {
        filesystem: {
          command: "npx",
          args: ["-y", "@modelcontextprotocol/server-filesystem", tempDir],
        },
      },
    },
    { codeMode: true }
  );

  const agent = new MCPAgent({
    llm: new ChatAnthropic({ model: ANTHROPIC_MODEL, temperature: 0 }),
    client,
    systemPrompt: PROMPTS.CODE_MODE,
    maxSteps: 30,
    autoInitialize: true,
  });

  try {
    for await (const _ of agent.prettyStreamEvents({
      prompt: "List all files in the workspace folder.",
    })) {
      // prettyStreamEvents prints formatted output
    }
  } finally {
    await agent.close();
    fs.rmSync(tempDir, { recursive: true, force: true });
  }
}

await main();
